README.md
# ProteoNeMo
This repository containes the code for pre-training and inference procedures of **protein language models** with [Nvidia NeMo toolkit](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/starthere/intro.html) from Peptone Ltd.

[![GitHub Super-Linter](https://github.com/peptoneinc/ProteoNeMo/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/marketplace/actions/super-linter)

ProteoNeMo can be used to extract residue level representations of proteins and to train related downstream tasks.

## Table of Contents

- [ProteoNeMo](#proteonemo)
  - [Table of Contents](#table-of-contents)
  - [Usage](#usage)
    - [Quick start](#quick-start)
    - [Datasets](#datasets)
    - [Download and preprocess datasets](#download-and-preprocess-datasets)
    - [ProteoNeMo pre-training](#proteonemo-pre-training)
    - [Residue level representations extraction](#residue-level-representations-extraction)
  - [Licence](#licence)

## Usage

### Quick-start

As a prerequisite, you must have [NeMo 1.7](https://github.com/NVIDIA/NeMo) or later installed to use this repository.

Install the **proteonemo** package:

Clone the ProteoNeMo repository, go to the ProteoNeMo directory and run

```bash
python setup.py install
```

### Datasets

ProteoNeMo can be pre-trained on:
* [UniRef](https://www.uniprot.org/uniref/)
  * UniRef 50
  * UniRef 90
  * UniRef 100
* [UniParc](https://www.uniprot.org/uniparc/)  
* [UniProtKB](https://www.uniprot.org/uniprot/)
  * UniProtKB Swiss-Prot
  * UniProtKB TrEMBL
  * [UniProtKB isoform sequences](https://www.uniprot.org/help/canonical_and_isoforms)

### Download and preprocess datasets

ProteoNeMo can be pre-trained on the datasets pointed-out above. You can choose your preferred one or make use of two or more of them at the same time.

Each dataset will be:
* **Downloaded** from [UniProt](https://www.uniprot.org/) and decopressed as a [.fasta](https://en.wikipedia.org/wiki/FASTA_format) file
* **Sharded** into several smaller `.txt` sub-files containing a random set of the related `.fasta` file, already splitted into **training** and **evaluation** samples
* **Tokenized** into several [.hdf5](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) files, one for each `.txt` sharded file, where the **masking** procedure has been already applied

In the ProteoNeMo directory run:
```bash
export BERT_PREP_WORKING_DIR=<your_dir>
cd scripts
bash create_datasets_from_start.sh <to_download> 
```

Where:

- `BERT_PREP_WORKING_DIR` defines the directory where the data will be downloaded and preprocessed
- `<to_download>` defines the datasets we want to download and preprocess where `uniref_50_only` is the default. 

The outputs are the `download`, `sharded` and `hdf5` directories under the `$BERT_PREP_WORKING_DIR` parent directory, containing the related files.

| To Download | Datasets |
|-------------|----------|
| `uniref_50_only` | **UniRef 50** |
| `uniref_all`| **UniRef 50, 90** and **100** |
| `uniparc`| **UniParc** |
| `uniprotkb_all`| **UniProtKB Swiss-Prot, TrEMBL** and  **isoform sequences** |

### ProteoNeMo pre-training

Once the download and preprocessing procedure is completed you're ready to pre-train ProteoNeMo.

The pre-training procedure exploits NeMo to solve the **Masked Language Modeling** (Masked LM) task. One training instance of Masked LM is a single modified protein sequence. Each token in the sentence has a 15% chance of being replaced by a [MASK] token. The chosen token is replaced with [MASK] 80% of the time, 10% with a random token and the remaining 10% the token is retained. The task is then to predict the original token.

We have currently integrated [BERT](https://arxiv.org/abs/1810.04805)-like *uncased* models from [HuggingFace](https://huggingface.co/).

The first thing you need to do is creating a `model_config.yaml` file in the [conf](conf) directory, specifying the relative pre-training and model options. You can use [this](conf/bert_pretrained_from_preprocessed_config.yaml) config as template. 

Take a look to [these](https://github.com/NVIDIA/NeMo/tree/main/tutorials) NeMo tutorials to get familiar with such options.

Secondly, you have to modify the `config_name` argument of the `@hydra_runner` decorator in [bert_pretraining.py](scripts/bert_pretraining.py)

Lastly, in the ProteoNeMo directory run:
```bash
cd scripts
python bert_pretraining.py 
```

The pre-training will start and a progress bar will appear &nbsp; ![](https://us-central1-progress-markdown.cloudfunctions.net/progress/50)

#### Tensorboard monitoring

Once the pre-training procedure has started a `nemo_experiments` directory will be automatically created under the [scripts](scripts) directory. 

Based on the `name: <PretrainingModelName>` parameter in the `.yaml` configuration file, a `<PretrainingModelName>` sub-directory containing all the related pre-training experiment logs will be created under `nemo_experiments`.

In the ProteoNeMo directory run: 
```bash
tensorboard --logdir=scripts/nemo_experiments/<PretrainingModelName> 
```

The Tensorboard UI will be available on port 6006

### Residue level representations extraction

Once a ProteoNeMo model will be pre-trained you'll get a `.nemo` file, placed in the `nemo_path` you've utilised in the `.yaml` configuration file.

You're now ready to extract the residue level representations of each protein a `.fasta` file.

In the ProteoNeMo directory run:
```bash
cd scripts
python bert_eval.py --input_file <fasta_input_file> \
                    --vocab_file ../static/vocab.txt \
                    --output_dir <reprs_output_dir> \
                    --model_file <nemo_pretrained_model>
```

Where:

- `--input_file` defines the `.fasta` file containing the proteins for which you want to extract the residue level representations
- `--vocab_file` defines the `.txt` file containing the vacabulary you want to use during the inference phase. We suggets you use the [standard](static/vocab.txt) one
- `--output_dir` defines the output directory where the residue level representations will be written. You'll get a `.pt` file for each protein sequence in the `--input_file` 
- `--model_file` defines the `.nemo` file used to get the pre-trained weights needed to get the residue level representations

## Licence

This source code is licensed under the Apache 2.0 license found in the `LICENSE` file in the root directory of this source tree.


CHANGELOG.md
# Changelog

## [v0.1.1](https://github.com/PeptoneInc/ProteoNeMo/tree/v0.1.1) (2022-03-14)

[Full Changelog](https://github.com/PeptoneInc/ProteoNeMo/compare/v0.1.0...v0.1.1)

## [v0.1.0](https://github.com/PeptoneInc/ProteoNeMo/tree/v0.1.0) (2022-03-07)

[Full Changelog](https://github.com/PeptoneInc/ProteoNeMo/compare/9156b2e0664b4fe26ae159acbd241e6ce16fef2a...v0.1.0)

**Merged pull requests:**

- Superlinter [\#1](https://github.com/PeptoneInc/ProteoNeMo/pull/1) ([CFisicaro](https://github.com/CFisicaro))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


scripts/bert_prep.py
# Copyright (c) 2021 Peptone.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from proteonemo.preprocessing.downloader import Downloader
from proteonemo.preprocessing.protein_sharding import Sharding

import argparse
import os
import pprint
import subprocess
import random


def main(args):
    working_dir = os.environ['BERT_PREP_WORKING_DIR']

    print('Working Directory:', working_dir)
    print('Action:', args.action)
    print('Dataset Name:', args.dataset)

    if args.input_files:
        args.input_files = args.input_files.split(',')

    hdf5_tfrecord_folder_prefix = "_seq_len_" + str(args.max_seq_length) \
                                  + "_max_pred_" + str(args.max_predictions_per_seq) + "_masked_lm_prob_" + str(args.masked_lm_prob) \
                                  + "_random_seed_" + str(args.random_seed) + "_dupe_factor_" + str(args.dupe_factor)

    directory_structure = {
        'download' : working_dir + '/download',    # Downloaded and decompressed
        'sharded' : working_dir + '/sharded_' + "training_shards_" + str(args.n_training_shards) + "_test_shards_" + str(args.n_test_shards) + "_fraction_" + str(args.fraction_test_set),
        'hdf5': working_dir + '/hdf5' + hdf5_tfrecord_folder_prefix
    }

    print('\nDirectory Structure:')
    pp = pprint.PrettyPrinter(indent=2)
    pp.pprint(directory_structure)
    print('')

    if args.action == 'download':
        if not os.path.exists(directory_structure['download']):
            os.makedirs(directory_structure['download'])

        downloader = Downloader(args.dataset, directory_structure['download'])
        downloader.download()

    elif args.action == 'sharding':
        # Note: uniref 50+90+100 requires user to provide list of input_files (comma-separated with no spaces)
        if args.dataset == 'uniparc' or 'uniref' in args.dataset or 'uniprotkb' in args.dataset:
            if args.input_files is None:
                if args.dataset == 'uniparc':
                    args.input_files = [directory_structure['download'] + '/uniparc/uniparc_active.fasta']
                elif args.dataset == 'uniref_50':
                    args.input_files = [directory_structure['download'] + '/uniref_50/uniref50.fasta']
                elif args.dataset == 'uniref_90':
                    args.input_files = [directory_structure['download'] + '/uniref_90/uniref90.fasta']
                elif args.dataset == 'uniref_100':
                    args.input_files = [directory_structure['download'] + '/uniref_100/uniref100.fasta']
                elif args.dataset == 'uniref_all':
                    args.input_files = [directory_structure['download'] + '/uniref_50/uniref50.fasta',
                        directory_structure['download'] + '/uniref_90/uniref90.fasta',
                        directory_structure['download'] + '/uniref_100/uniref100.fasta']
                elif args.dataset == 'uniprotkb_swissprot':
                    args.input_files = [directory_structure['download'] + '/uniprotkb_swissprot/uniprot_sprot.fasta']
                elif args.dataset == 'uniprotkb_trembl':
                    args.input_files = [directory_structure['download'] + '/uniprotkb_trembl/uniprot_trembl.fasta']
                elif args.dataset == 'uniprotkb_isoformseqs':
                    args.input_files = [directory_structure['download'] + '/uniprotkb_isoformseqs/uniprot_sprot_varsplic.fasta']
                elif args.dataset == 'uniprotkb_all':
                    args.input_files = [directory_structure['download'] + '/uniprotkb_swissprot/uniprot_sprot.fasta',
                        directory_structure['download'] + '/uniprotkb_trembl/uniprot_trembl.fasta',
                        directory_structure['download'] + '/uniprotkb_isoformseqs/uniprot_sprot_varsplic.fasta']

            output_file_prefix = directory_structure['sharded'] + '/' + args.dataset + '/' + args.dataset

            if not os.path.exists(directory_structure['sharded']):
                os.makedirs(directory_structure['sharded'])

            if not os.path.exists(directory_structure['sharded'] + '/' + args.dataset):
                os.makedirs(directory_structure['sharded'] + '/' + args.dataset)

            rng = random.Random(args.random_seed)
            sharding = Sharding(args.input_files, output_file_prefix, args.n_training_shards, args.n_test_shards, args.fraction_test_set, rng)
            sharding.load_fastas()
            sharding.write_shards_to_disk()

        else:
            assert False, 'Unsupported dataset for sharding'

    elif args.action == 'create_hdf5_files':
        last_process = None

        if not os.path.exists(directory_structure['hdf5'] + "/" + args.dataset):
            os.makedirs(directory_structure['hdf5'] + "/" + args.dataset)

        def create_record_worker(filename_prefix, shard_id, output_format='hdf5'):
            bert_preprocessing_command = 'python ../proteonemo/preprocessing/create_pretraining_data.py'
            bert_preprocessing_command += ' --input_file=' + directory_structure['sharded'] + '/' + args.dataset + '/' + filename_prefix + '_' + str(shard_id) + '.txt'
            bert_preprocessing_command += ' --output_file=' + directory_structure['hdf5'] + '/' + args.dataset + '/' + filename_prefix + '_' + str(shard_id) + '.' + output_format
            bert_preprocessing_command += ' --vocab_file=' + args.vocab_file
            bert_preprocessing_command += ' --small_vocab_file=' + args.small_vocab_file
            bert_preprocessing_command += ' --do_upper_case' if args.do_upper_case else ''
            bert_preprocessing_command += ' --max_seq_length=' + str(args.max_seq_length)
            bert_preprocessing_command += ' --max_predictions_per_seq=' + str(args.max_predictions_per_seq)
            bert_preprocessing_command += ' --masked_lm_prob=' + str(args.masked_lm_prob)
            bert_preprocessing_command += ' --random_seed=' + str(args.random_seed)
            bert_preprocessing_command += ' --dupe_factor=' + str(args.dupe_factor)
            bert_preprocessing_process = subprocess.Popen(bert_preprocessing_command, shell=True)

            last_process = bert_preprocessing_process

            # This could be better optimized (fine if all take equal time)
            if shard_id % args.n_processes == 0 and shard_id > 0:
                bert_preprocessing_process.wait()
            return last_process

        output_file_prefix = args.dataset

        for i in range(args.n_training_shards):
            last_process = create_record_worker(output_file_prefix + '_training', i)

        last_process.wait()

        for i in range(args.n_test_shards):
            last_process = create_record_worker(output_file_prefix + '_test', i)

        last_process.wait()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description='Preprocessing Application for Everything BERT-related'
    )

    parser.add_argument(
        '--action',
        type=str,
        help='Specify the action you want the app to take. e.g., download data, or create hdf5 files',
        choices={
            'download',               # Download and verify mdf5/sha sums
            'sharding',               # Convert previous formatted text into shards containing one sentence per line
            'create_hdf5_files'       # Turn each shard into a HDF5 file with masking and next sentence prediction info
        }
    )

    parser.add_argument(
        '--dataset',
        type=str,
        help='Specify the dataset to perform --action on',
        choices={
            'uniref_50',
            'uniref_90',
            'uniref_100',
            'uniref_all',
            'uniprotkb_swissprot',
            'uniprotkb_trembl',
            'uniprotkb_isoformseqs',
            'uniprotkb_all',
            'uniparc',
            'all'
        }
    )

    parser.add_argument(
        '--input_files',
        type=str,
        help='Specify the input files in a comma-separated list (no spaces)'
    )

    parser.add_argument(
        '--n_training_shards',
        type=int,
        help='Specify the number of training shards to generate',
        default=256
    )

    parser.add_argument(
        '--n_test_shards',
        type=int,
        help='Specify the number of test shards to generate',
        default=256
    )

    parser.add_argument(
        '--fraction_test_set',
        type=float,
        help='Specify the fraction (0..1) of the data to withhold for the test data split (based on number of sequences)',
        default=0.1
    )

    parser.add_argument(
        '--n_processes',
        type=int,
        help='Specify the max number of processes to allow at one time',
        default=4
    )

    parser.add_argument(
        '--random_seed',
        type=int,
        help='Specify the base seed to use for any random number generation',
        default=12345
    )

    parser.add_argument(
        '--dupe_factor',
        type=int,
        help='Specify the duplication factor',
        default=5
    )

    parser.add_argument(
        '--masked_lm_prob',
        type=float,
        help='Specify the probability for masked lm',
        default=0.15
    )

    parser.add_argument(
        '--max_seq_length',
        type=int,
        help='Specify the maximum sequence length',
        default=1024
    )

    parser.add_argument(
        '--max_predictions_per_seq',
        type=int,
        help='Specify the maximum number of masked words per sequence',
        default=160
    )

    parser.add_argument(
        '--vocab_file',
        type=str,
        help='Specify absolute path to vocab file to use)'
    )

    parser.add_argument(
        '--small_vocab_file',
        type=str,
        help='Specify absolute path to vocab file to use)'
    )

    parser.add_argument("--do_upper_case",
        action='store_true',
        default=True,
        help="Whether to upper case the input text. True for uncased models, False for cased models.")

    args = parser.parse_args()
    main(args)
